File: src/main/java/be/superjoran/mint/domain/DomainObject.java
package be.superjoran.mint.domain;

import org.hibernate.annotations.GenericGenerator;

import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.MappedSuperclass;
import java.io.Serializable;

/**
 * Created by Jorandeboever
 * Date: 01-Oct-16.
 */
@MappedSuperclass
public abstract class DomainObject implements Serializable, Cloneable, Display {
    private static final long serialVersionUID = -7793260716425423013L;

    @Id
    @GeneratedValue(generator = "uuid2")
    @GenericGenerator(name = "uuid2", strategy = "uuid2")
    private String uuid;

    public String getUuid() {
        return this.uuid;
    }

    public void setUuid(String uuid) {
        this.uuid = uuid;
    }

    @Override
    protected Object clone() throws CloneNotSupportedException {
        DomainObject clone = (DomainObject) super.clone();
        clone.uuid = this.uuid;
        return clone;
    }

    @Override
    public String getDisplayValue(){
        return this.getUuid();
    }

    @Override
    public String getId() {
        return this.getUuid();
    }

    @Override
    public boolean equals(Object obj) {
        if(obj instanceof DomainObject){
            return this.getUuid().equals(((DomainObject) obj).getUuid());
        }
        return super.equals(obj);
    }
}

File: src/main/java/be/superjoran/mint/domain/DomainObject.kt
package be.superjoran.mint.domain

import org.hibernate.annotations.GenericGenerator
import java.io.Serializable
import javax.persistence.GeneratedValue
import javax.persistence.Id
import javax.persistence.MappedSuperclass

/**
 * Created by Jorandeboever
 * Date: 01-Oct-16.
 */
@MappedSuperclass
abstract class DomainObject : Serializable, Cloneable, Display {

    @Id
    @GeneratedValue(generator = "uuid2")
    @GenericGenerator(name = "uuid2", strategy = "uuid2")
    var uuid: String? = null

    constructor()

    constructor(uuid: String) {
        this.uuid = uuid
    }

    override fun equals(other: Any?): Boolean {
        if (other is DomainObject) {
            return other.uuid == this.uuid
        }
        return super.equals(other)
    }

    override fun hashCode(): Int {
        return this.uuid?.hashCode() ?: super.hashCode()
    }

    override val id: String?
        get() = this.uuid
    override val displayValue: String?
        get() = this.toString()

}
